### DISCUSSION TITLE
Thoughts on std::colony as replacement for sparse_set implementation?

### DISCUSSION THREAD

**Discussion by jonesmz** [skypjack/entt#663] [Ideas]
Just curious if folks know about this standard addition proposal, and whether or not it would be a fit for entt, if it's adopted.

https://plflib.org/colony.htm

**Comment by skypjack** [comment]
I know of `plf::colony`, yeah. Very interesting, even though I don't see where it could be a fit for EnTT. Can you elaborate a little further?
By the way, I didn't know there was a proposal for integrating it into the standard. Any ETA or feedback from the committee already?

**Reply by jonesmz** [reply]
> Can you elaborate a little further?

I ask out of ignorance of the internals of EnTT, as I have not yet had time to study the entire implementation, only some of it.

From high level perspective of an outsider, the std/plf::colony claims of memory stability until a re-arrangement is trigged, is attractive. In the project that I'm working on that uses EnTT (The Open Space Program game, which you've commented on a time or two), there are some areas of the code where it would be somewhat more convenient if we had memory stability guarantees.

But as I said, I don't know (yet) anything about how the actual storage of EnTT is implemented, so it may only be a question that makes sense to ask because I haven't dug deep enough yet.

> Any ETA or feedback from the committee already?

Not that I am aware of.

**Reply by skypjack** [reply]
I see, no worries. EnTT uses a model with independent pools. Each of them is a sparse set. The actual storage is in sync with the packed array of the sparse set. Morover, custom pools are supported, so you can mess with this stack as you like. 🙂 
Currently, only the sparse array is paged and that's why you don't have reference stability. There is a [feature request](https://github.com/skypjack/entt/issues/625) for that and a _plan_ to introduce two models: weak references that refresh on use and truly stable references. Actually, my last answer to that discussion is a little outdated and things changed in the meantime, but still. This should happen soon, because I need it for work too.
`plf::colony` is really amazing, though probably not a good replacement for the sparse set. Its characteristics seem to favor a lot linear accesses but to suffer from random accesses, while a sparse set shines at both (and has its own cons somewhere else ofc).
